src/main/java/frc/robot/subsystems/SwerveModule.java

```diff
@@ -0,0 +1,234 @@
+
+        this.printName = String.format("SwerveModule[%s]", identifier);
+
+        this.drive.restoreFactoryDefaults();
+        this.steer.restoreFactoryDefaults();
+
+        this.driveEncoder = this.drive.getEncoder();
+        this.steerEncoder = this.steer.getEncoder();
+        this.drivePID = this.drive.getPIDController();
+        this.steerPID = this.steer.getPIDController();
+
+        driveEncoder.setPosition(0);
+        steerEncoder.setPosition(0);
+
+        drivePID.setP(DRIVE_P);
+        drivePID.setI(DRIVE_I);
+        drivePID.setD(DRIVE_D);
+        drivePID.setFF(DRIVE_F);
+        drivePID.setOutputRange(-1, 1);
+
+        steerPID.setP(STEER_P);
+        steerPID.setI(STEER_I);
+        steerPID.setD(STEER_D);
+        steerPID.setFF(STEER_F);
+        steerPID.setOutputRange(-1, 1);
+        steerPID.setPositionPIDWrappingMaxInput(STEER_GEAR_RATIO);
+        steerPID.setPositionPIDWrappingMinInput(0);
+        steerPID.setPositionPIDWrappingEnabled(true);
+
+        steerEncoder.setPosition((this.canCoder.getAbsolutePosition().refresh().getValue() - zeroAngle) / 360 * STEER_GEAR_RATIO);
+    }
+
+    public double getHeadingDegrees() {
+        double degrees = steerEncoder.getPosition() / STEER_GEAR_RATIO * 360;
+        degrees %= 360;
+        if (degrees < 0) {
+            degrees += 360;
+        }
+
+        return degrees;
+    }
+
+    private double getAbsoluteHeadingDegrees() {
+        return canCoder.getAbsolutePosition().refresh().getValue();
+    }
+
+    public double getVelocityRpm(){
+        return driveEncoder.getVelocity() / DRIVE_GEAR_RATIO;
+    }
+
+    public double getVelocityMetersPerSecond() {
+        return getVelocityRpm() / 60 * WHEEL_CIRCUMFERENCE;
+    }
+
+    public double getPositionMeters() {
+        return driveEncoder.getPosition() / DRIVE_GEAR_RATIO * WHEEL_CIRCUMFERENCE;
+    }
+
+    public Rotation2d getRotation() {
+        return Rotation2d.fromDegrees(getHeadingDegrees());
+    }
+
+    public Measure<Voltage> getOutputVoltage() {
+        double voltage = drive.getBusVoltage() * drive.getAppliedOutput();
+        return Volts.of(voltage);
+    }
+
+    public Measure<Velocity<Distance>> getLinearVelocity() {
+        double velocity = getVelocityMetersPerSecond();
+        return MetersPerSecond.of(velocity);
+    }
+
+    public SwerveModuleState getModuleStates(){
+        return new SwerveModuleState(
+                getVelocityMetersPerSecond(),
+                getRotation()
+        );
+    }
+
+    public SwerveModulePosition getModulePosition(){
+        return new SwerveModulePosition(
+                getPositionMeters(),
+                getRotation()
+        );
+    }
+
+    public void resetEncoders() {
+        steerEncoder.setPosition(0);
+        driveEncoder.setPosition(0);
+    }
+
+    public void setDesiredState(SwerveModuleState desiredState) {
+        SwerveModuleState state = SwerveModuleState.optimize(desiredState,getRotation());
```

Comment:
you tested this on the floor? does the swerve handle well with different angles of motion? did you notice if the steer is fluid or does it go a bit too much?

If it's fine then delete the old `optimize` in this class